Make Stream.stop() async

uvjustin opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
`Stream.stop()` calls `Thread.join()` which is blocking. This PR makes `Stream.stop()` async and uses an executor job to wrap the blocking part of the method.
This PR also makes stream's shutdown handler async, and the handler now calls stop() concurrently across all the workers. This may speed up shutdown, although currently the blocking time should be capped by the time it takes to receive another packet which should be minimal.
 
## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #71876
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

uvjustin committed:
Make Stream.stop() async

uvjustin committed:
Adjust nest usage

MartinHjelmare commented on homeassistant/components/stream/__init__.py:
```diff
@@ -272,11 +276,11 @@ def add_provider(
                     not self.keepalive or fmt == RECORDER_PROVIDER
                 ) and fmt in self._outputs:
-                    self.remove_provider(self._outputs[fmt])
+                    hass.add_job(self.remove_provider, self._outputs[fmt])
```
We should use the async api inside async context.
```suggestion
                    hass.async_create_task(self.remove_provider, self._outputs[fmt])
```

uvjustin replied:
This change seems to be causing the tests to fail with `RuntimeError: Non-thread-safe operation invoked on an event loop other than the current one`, but I'm not sure why. This idle_callback gets called from within async_call_later which should be in the main event loop, no?

MartinHjelmare replied:
The fire method isn't decorated with `@callback`,so it will be scheduled on the executor.

https://github.com/home-assistant/core/blob/e5b447839af38d5c10e3205ef095aa205cb6dd16/homeassistant/components/stream/core.py#L222

uvjustin replied:
Good catch, thanks!

MartinHjelmare commented on homeassistant/components/stream/__init__.py:
```diff
@@ -394,25 +398,24 @@ def _run_worker(self) -> None:
                 self._async_update_state(True)
             for provider in self.outputs().values():
-                self.remove_provider(provider)
+                await self.remove_provider(provider)
```
Can we gather this work?

uvjustin replied:
I don't think that it's necessary here since the only thing that is blocking is the worker close, and that will only happen on removing the last provider.

MartinHjelmare replied:
Maybe add a comment why we don't gather the work?

uvjustin replied:
Yes, makes sense, will do that now

uvjustin committed:
Fix nest test

uvjustin committed:
Use async_create_task

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

uvjustin committed:
Add comment

uvjustin committed:
Fix test

uvjustin committed:
Decorate IdleTimer.fire with @callback

allenporter commented on homeassistant/components/stream/__init__.py:
```diff
@@ -272,11 +276,11 @@ def add_provider(
                     not self.keepalive or fmt == RECORDER_PROVIDER
                 ) and fmt in self._outputs:
-                    self.remove_provider(self._outputs[fmt])
+                    hass.async_create_task(self.remove_provider(self._outputs[fmt]))
```
I believe check idle needs to have the output removed so this removal cannot be deferred

allenporter replied:
I'm not sure if that implies the idle timer must also be async or if the stop call in remove should be the task instead, or it returns a future for the caller to decide after it does the removal.

uvjustin replied:
Can probably just make the callback async and call it from fire

allenporter commented on homeassistant/components/stream/core.py:
```diff
@@ -219,11 +222,12 @@ def clear(self) -> None:
         self.idle = True
         self._unsub = None
-        self._callback()
+        self._callback(self._hass)
```
I think the idle function can access hass on the stream object (or as a local copy) to avoid having to plumb this through

allenporter reviewed (commented):
Reviewed on mobile for a fast initial pass but apologies if the mobile view means I missed some details.

uvjustin committed:
Make idle_callback async

allenporter commented:
Now that `_stop` runs in another thread, I worry that its mutating `_thread` which used to only happen in the event loop possibly adding some raciness for start/stop called in parallel. Maybe reseting the thread needs to be done in the async loop, but the waiting for the join still can be done to the other thread?

uvjustin committed:
Only call join() in executor

uvjustin commented:
> Now that `_stop` runs in another thread, I worry that its mutating `_thread` which used to only happen in the event loop possibly adding some raciness for start/stop called in parallel. Maybe reseting the thread needs to be done in the async loop, but the waiting for the join still can be done to the other thread?

I changed it to only use the executor for the join() call. It looks like touching Stream._thread is now only done from the event loop.
I moved setting Stream._thread to None to above the join to keep it contiguous without a possible async context switch in the middle. However look at it now I don't think that's right either - we want to keep from setting the thread to None until the prior thread is completely finished to avoid having two worker threads going at once.
It seems to keep it simplest we would like the _stop() function to be atomic at least wrt to the start() function. How about using an asyncio lock?

uvjustin committed:
Set thread to None after join

uvjustin committed:
Make start async
Use asyncio.Lock for start and _stop

uvjustin committed:
Change start() to await start() in camera

uvjustin committed:
Use AsyncMock for Stream.start in nest tests

allenporter reviewed (approved):
That seems like a good solution. I also liked the intermediate change where `_stop` had the await for join only as it seemed correct. However, I remember the cases with start/stop up in camera being racy -- this solution with the lock is growing on me since it seems the most explicit. There is enough subtle stuff going on in stream in general, so with your change here we don't to worry about it as much anymore.  👍🏼

MartinHjelmare commented on homeassistant/components/stream/__init__.py:
```diff
@@ -316,22 +320,26 @@ def _async_update_state(self, available: bool) -> None:
+        Uses an asyncio.Lock to avoid conflicts with _stop().
+        """
+        async with self._start_stop_lock:
+            if self._thread is None or not self._thread.is_alive():
```
Please invert the check and return early if true. Then we can outdent below.

MartinHjelmare commented on homeassistant/components/stream/__init__.py:
```diff
@@ -394,35 +402,40 @@ def _run_worker(self) -> None:
+        Uses an asyncio.Lock to avoid conflicts with start().
+        """
+        async with self._start_stop_lock:
+            if self._thread is not None:
```
As above.

MartinHjelmare commented on homeassistant/components/stream/__init__.py:
```diff
@@ -394,35 +402,40 @@ def _run_worker(self) -> None:
+                self._thread_quit.set()
+                await self.hass.async_add_executor_job(self._thread.join)
+                self._thread = None
+                self._logger.info(
```
We normally use debug level for all integration messages that aren't warnings or errors.

MartinHjelmare commented on homeassistant/components/stream/__init__.py:
```diff
@@ -394,35 +402,40 @@ def _run_worker(self) -> None:
-    def stop(self) -> None:
-        """Remove outputs and access token."""
+    async def stop(self) -> None:
+        """Remove outputs and access token. Should only be called from the event loop."""
```
It's a coroutine function. We don't need to say that it should run in the event loop.
```suggestion
        """Remove outputs and access token."""
```

MartinHjelmare commented on homeassistant/components/stream/__init__.py:
```diff
@@ -394,35 +402,40 @@ def _run_worker(self) -> None:
+                await self.remove_provider(provider)
 
-        self.hass.loop.call_soon_threadsafe(worker_finished)
+        self.hass.add_job(worker_finished)
```
If we know the target is a coroutine we can use `hass.create_task` from another thread.
```suggestion
        self.hass.create_task(worker_finished())
```

uvjustin replied:
Thanks, made all the changes suggested here.

uvjustin committed:
Return early
Change info messages to debug
Remove unnecessary annotation in docstring
Use create_task instead of add_job

MartinHjelmare reviewed (approved):
Looks good!

allenporter reviewed (approved):
Nice, looks good.

uvjustin committed:
Use list instead of set for awaitables

uvjustin commented:
Thanks for the reviews. Just made a tiny change (list vs set comprehension) and merged it.

allenporter reviewed (commented):
👍🏼 looks good
